File: app/src/main/java/com/example/buildmypc/ui/parts/parts/Storage.java
package com.example.buildmypc.ui.parts.parts;

import android.os.Parcel;
import android.os.Parcelable;

import org.jetbrains.annotations.NotNull;

public class Storage extends Part implements Parcelable {
	private double formFactor;
	private int cacheSizeMB; // in mb
	private String capacity; // in MB -> 2TB is 2000, 1.5TB is 1500, 512 GB is 512
	private String sataInterface; // which SATA does this part connect to
	private boolean nvme; // NMVe, or non-volatile memory express
	private int rpm; // only applies for HDDs
	private String type; // either HDD or SSD

	public Storage(String model, String manufacturer) {
		super(model, manufacturer);
	}

	public Storage(String model, String manufacturer, double formFactor, int cacheSizeMB, String capacity, String hardwareInterface, boolean nvme, int rpm, String type) {
		super(model, manufacturer);
		this.formFactor = formFactor;
		this.cacheSizeMB = cacheSizeMB;
		this.capacity = capacity;
		this.sataInterface = hardwareInterface;
		this.nvme = nvme;
		this.rpm = rpm;
		this.type = type;
	}

	public Storage(Parcel in) {
		super(in.toString(), in.toString());
		formFactor = in.readDouble();
		cacheSizeMB = in.readInt();
		capacity = in.readString();
		sataInterface = in.readString();
		nvme = in.readBoolean();
		rpm = in.readInt();
		type = in.readString();
	}

	public int getCacheSizeMB() {
		return cacheSizeMB;
	}

	public void setCacheSizeMB(int cacheSizeMB) {
		this.cacheSizeMB = cacheSizeMB;
	}

	public String  getCapacity() {
		return capacity;
	}

	public void setCapacity(String capacity) {
		this.capacity = capacity;
	}

	public double getFormFactor() {
		return formFactor;
	}

	public void setFormFactor(double formFactor) {
		this.formFactor = formFactor;
	}

	public String getSataInterface() {
		return sataInterface;
	}

	public void setSataInterface(String sataInterface) {
		this.sataInterface = sataInterface;
	}

	public boolean getNvme() {
		return nvme;
	}

	public void setNvme(boolean nvme) {
		this.nvme = nvme;
	}

	public int getRpm() {
		return rpm;
	}

	public void setRpm(int rpm) {
		this.rpm = rpm;
	}

	public String getType() {
		return type;
	}

	public void setType(String type) {
		this.type = type;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		if (!super.equals(o)) return false;

		Storage storage = (Storage) o;

		if (Double.compare(storage.getFormFactor(), getFormFactor()) != 0) return false;
		if (getCacheSizeMB() != storage.getCacheSizeMB()) return false;
		if (getNvme() != storage.getNvme()) return false;
		if (getRpm() != storage.getRpm()) return false;
		if (getCapacity() != null ? !getCapacity().equals(storage.getCapacity()) : storage.getCapacity() != null)
			return false;
		if (getSataInterface() != null ? !getSataInterface().equals(storage.getSataInterface()) : storage.getSataInterface() != null)
			return false;
		return getType() != null ? getType().equals(storage.getType()) : storage.getType() == null;
	}

	@Override
	public int hashCode() {
		int result = super.hashCode();
		long temp;
		temp = Double.doubleToLongBits(getFormFactor());
		result = 31 * result + (int) (temp ^ (temp >>> 32));
		result = 31 * result + getCacheSizeMB();
		result = 31 * result + (getCapacity() != null ? getCapacity().hashCode() : 0);
		result = 31 * result + (getSataInterface() != null ? getSataInterface().hashCode() : 0);
		result = 31 * result + (getNvme() ? 1 : 0);
		result = 31 * result + getRpm();
		result = 31 * result + (getType() != null ? getType().hashCode() : 0);
		return result;
	}

	@Override
	public void writeToParcel(@NotNull Parcel dest, int flags) {
		super.writeToParcel(dest, flags);
		dest.writeDouble(formFactor);
		dest.writeInt(cacheSizeMB);
		dest.writeString(capacity);
		dest.writeString(sataInterface);
		dest.writeBoolean(nvme);
		dest.writeInt(rpm);
		dest.writeString(type);
	}
}

File: app/src/main/java/com/example/buildmypc/ui/parts/parts/Storage.kt
package com.example.buildmypc.ui.parts.parts

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import org.jetbrains.annotations.Contract

class Storage : Part, Parcelable {
	var formFactor: String? = null
	var cacheSizeMB // in mb
			= 0
	var capacity // in MB -> 2TB is 2000, 1.5TB is 1500, 512 GB is 512
			: String? = null
	var sataInterface // which SATA does this part connect to
			: String? = null
	var nvme // NMVe, or non-volatile memory express
			= false
	var rpm // only applies for HDDs
			= 0
	var type // either HDD or SSD
			: String? = null

	constructor(model: String?, manufacturer: String?) : super(model, manufacturer)
	constructor(
		model: String?,
		manufacturer: String?,
		formFactor: String?,
		cacheSizeMB: Int,
		capacity: String?,
		hardwareInterface: String?,
		nvme: Boolean,
		rpm: Int,
		type: String?
	) : super(model, manufacturer) {
		this.formFactor = formFactor
		this.cacheSizeMB = cacheSizeMB
		this.capacity = capacity
		this.sataInterface = hardwareInterface
		this.nvme = nvme
		this.rpm = rpm
		this.type = type
	}

	constructor(`in`: Parcel) : super(`in`.toString(), `in`.toString()) {
		formFactor = `in`.readString()
		cacheSizeMB = `in`.readInt()
		capacity = `in`.readString()
		sataInterface = `in`.readString()
		nvme = `in`.readBoolean()
		rpm = `in`.readInt()
		type = `in`.readString()
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other == null || javaClass != other.javaClass) return false
		if (!super.equals(other)) return false
		val storage = other as Storage
		if (cacheSizeMB != storage.cacheSizeMB) return false
		if (nvme != storage.nvme) return false
		if (rpm != storage.rpm) return false
		if (if (formFactor != null) formFactor != storage.formFactor else storage.formFactor != null) return false
		if (if (capacity != null) capacity != storage.capacity else storage.capacity != null) return false
		if (if (sataInterface != null) sataInterface != storage.sataInterface else storage.sataInterface != null) return false
		return if (type != null) type == storage.type else storage.type == null
	}

	override fun hashCode(): Int {
		var result = super.hashCode()
		result = 31 * result + if (formFactor != null) formFactor.hashCode() else 0
		result = 31 * result + cacheSizeMB
		result = 31 * result + if (capacity != null) capacity.hashCode() else 0
		result = 31 * result + if (sataInterface != null) sataInterface.hashCode() else 0
		result = 31 * result + if (nvme) 1 else 0
		result = 31 * result + rpm
		result = 31 * result + if (type != null) type.hashCode() else 0
		return result
	}

	override fun writeToParcel(dest: Parcel, flags: Int) {
		super.writeToParcel(dest, flags)
		dest.writeString(formFactor)
		dest.writeInt(cacheSizeMB)
		dest.writeString(capacity)
		dest.writeString(sataInterface)
		dest.writeBoolean(nvme)
		dest.writeInt(rpm)
		dest.writeString(type)
	}

	override val paramCount = 9

	companion object {
		@JvmField
		val CREATOR: Creator<Storage> = object : Creator<Storage> {
			@Contract("_ -> new")
			override fun createFromParcel(`in`: Parcel) = Storage(`in`)

			@Contract(value = "_ -> new", pure = true)
			override fun newArray(size: Int): Array<Storage?> = arrayOfNulls(size)
		}
	}
}